### PR TITLE
SHA-1 and SHA-256 x86

### PR DESCRIPTION
**Pull request by MarekKnapek**

### CODE REVIEW

File: Userland/Libraries/LibCrypto/Hash/SHA1.cpp [hunk 1]
```diff
@@ -7,18 +7,34 @@
...
+#include <AK/Platform.h>
 #include <AK/Types.h>
 #include <LibCrypto/Hash/SHA1.h>
 
+#if ARCH(I386) || ARCH(X86_64)
+#    include <emmintrin.h> // SSE2 _mm_add_epi32 _mm_loadu_si128 _mm_set_epi32 _mm_set_epi64x _mm_setzero_si128 _mm_shuffle_epi32 _mm_storeu_si128 _mm_xor_si128
+#    include <tmmintrin.h> // SSSE3 _mm_shuffle_epi8
+#    include <smmintrin.h> // SSE4.1 _mm_extract_epi32
+#    include <immintrin.h> // SHA _mm_sha1msg1_epu32 _mm_sha1msg2_epu32 _mm_sha1nexte_epu32 _mm_sha1rnds4_epu32
+#    define sha1_attribute_target_default __attribute__((target("default")))
```
Comment (ADKaster) [inline Userland/Libraries/LibCrypto/Hash/SHA1.cpp:19]: macros should be SCREAMING_SNAKE_CASE

File: Userland/Libraries/LibCrypto/Hash/SHA1.cpp [hunk 2]
```diff
@@ -7,18 +7,34 @@
...
+#    define sha1_attribute_target_x86 __attribute__((target("sse2,ssse3,sse4.1,sha")))
+#else
+#    define sha1_attribute_target_default
+#    define sha1_attribute_target_x86
```
Comment (ADKaster) [inline Userland/Libraries/LibCrypto/Hash/SHA1.cpp:23]: don't need this for non-x86, it's not referenced

File: Userland/Libraries/LibCrypto/Hash/SHA1.cpp [hunk 3]
```diff
@@ -7,18 +7,34 @@
...
+
 namespace Crypto::Hash {
 
 static constexpr auto ROTATE_LEFT(u32 value, size_t bits)
 {
     return (value << bits) | (value >> (32 - bits));
 }
 
-inline void SHA1::transform(u8 const* data)
+sha1_attribute_target_default static void transform_impl(u32* state, u8 const* data)
```
Comment (ADKaster) [inline Userland/Libraries/LibCrypto/Hash/SHA1.cpp:34]: can we use Spans or Array& or something instead of raw pointers with no size information? This is partially pre-existing in this file but there's no reason to propagate that mistake further.

File: Userland/Libraries/LibCrypto/Hash/SHA1.cpp [hunk 4]
```diff
@@ -67,6 +83,188 @@ inline void SHA1::transform(u8 const* data)
...
+    __m128i reverse_8;
+    __m128i abcd_1;
+    __m128i e;
+    __m128i old_abcd;
+    __m128i old_e;
+    __m128i msg_0;
+    __m128i abcd_2;
+    __m128i msg_1;
+    __m128i msg_2;
+    __m128i msg_3;
```
Comment (ADKaster) [inline Userland/Libraries/LibCrypto/Hash/SHA1.cpp:98]: This is still very C.

1) Initialize all primitive type variables. 
2) Only define variables at first use, not at the top of the function. We aren't in C89 anymore

File: Userland/Libraries/LibCrypto/Hash/SHA1.cpp [hunk 5]
```diff
@@ -67,6 +83,188 @@ inline void SHA1::transform(u8 const* data)
...
+
+    VERIFY(state);
+    VERIFY(data);
+
+    // Set up constant for reversing input buffer.
+    reverse_8 = _mm_set_epi64x(0x0001020304050607ull, 0x08090a0b0c0d0e0full);
+    // Load state into working registers.
+    abcd_1 = _mm_loadu_si128(reinterpret_cast<__m128i const*>(&state[0]));
```
Comment (ADKaster) [inline Userland/Libraries/LibCrypto/Hash/SHA1.cpp:106]: Does all of this *really* need to be 100% unrolled? It's very hard to follow, and uses all these temporary variables in a very unintuitve way.

Comment (ADKaster) [reply Userland/Libraries/LibCrypto/Hash/SHA1.cpp:106]: Also what are 'abcd' and 'e'? Are they names from the spec? or just arbitrary temp variable names? Descriptive names are much preferred.

File: Userland/Libraries/LibCrypto/Hash/SHA1.cpp (hunk 4 above)
Comment (ADKaster) [inline Userland/Libraries/LibCrypto/Hash/SHA1.cpp:98]: Any reason not to put some of these variables in an array, or even better, an AK::Array?

File: Userland/Libraries/LibCrypto/Hash/SHA1.cpp [hunk 6]
```diff
@@ -67,6 +83,188 @@ inline void SHA1::transform(u8 const* data)
...
+    __m128i old_abcd;
+    __m128i old_e;
+    __m128i msg_0;
+    __m128i abcd_2;
+    __m128i msg_1;
+    __m128i msg_2;
+    __m128i msg_3;
+
+    VERIFY(state);
+    VERIFY(data);
```
Comment (ADKaster) [inline Userland/Libraries/LibCrypto/Hash/SHA1.cpp:101]: If we used Spans or references to arrays, these verifies go way.

File: Userland/Libraries/LibCrypto/Hash/SHA1.cpp [hunk 7]
```diff
@@ -67,6 +83,188 @@ inline void SHA1::transform(u8 const* data)
...
+    msg_0 = _mm_setzero_si128();
+    e = _mm_sha1nexte_epu32(abcd_2, msg_0);
+    // Sum current working registers with saved state from before.
+    abcd_1 = _mm_add_epi32(abcd_1, old_abcd);
+    e = _mm_add_epi32(e, old_e);
+    // Here could end `for' or `while' loop in case we processed more than one block at once.
+    // Save working registers into state.
+    abcd_1 = _mm_shuffle_epi32(abcd_1, 0x1b);
+    _mm_storeu_si128(reinterpret_cast<__m128i*>(&state[0]), abcd_1);
+    *reinterpret_cast<int*>(&state[4]) = _mm_extract_epi32(e, 3);
```
Comment (ADKaster) [inline Userland/Libraries/LibCrypto/Hash/SHA1.cpp:259]: This is cast and storage is not safe, for alignment reasons.

Can we mark alignment requirements on the input state variable? If it's a member variable of the SHAXX class, we should absolutely set an alignment requirement on it, to avoid hard to debug trap/crashes.

We have AK::ByteReader for this.

File: Userland/Libraries/LibCrypto/Hash/SHA2.cpp [hunk 8]
```diff
@@ -34,16 +47,16 @@ inline void SHA256::transform(u8 const* data)
         m[i] = (data[j] << 24) | (data[j + 1] << 16) | (data[j + 2] << 8) | data[j + 3];
     }
 
-    for (; i < BlockSize; ++i) {
+    for (; i < 64; ++i) {
```
Comment (ADKaster) [inline Userland/Libraries/LibCrypto/Hash/SHA2.cpp:50]: can we not use SHA256::BlockSize here? replacing a meaningfully named constant with a magic number is less than ideal.

**Review by ADKaster — COMMENTED** [review]
Seems neat, but is missing a few things:

- performance numbers from some benchmark
- if possible, a godbolt.org link proving that the intrinsic-heavy code needs to have the loops completely unrolled. (i.e., that the compiler doesn't generate the same or better code with a legible for loop).

**Comment by MarekKnapek** [comment]
> * performance numbers from some benchmark

OK, I will work on that. But all my computers are virtual computers (non-repeatable performance). The one that can build Serenity (Linux) does not support the SHA instructions the other (Windows without WSL) can not run the code from Serenity. Will create benchmark that mimics Serenity code as close as possible. But those numbers would be pulled out of my a**.

> * if possible, a godbolt.org link proving that the intrinsic-heavy code needs to have the loops completely unrolled. (i.e., that the compiler doesn't generate the same or better code with a legible for loop).

The code needs to be unrolled, because of code such as `abcd_2 = _mm_sha1rnds4_epu32(abcd_1, e, 3);` in the SHA-1. The [intrinsic](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=_mm_sha1rnds4_epu32&ig_expand=5877) does take `imm8` not a variable.

File: Userland/Libraries/LibCrypto/Hash/SHA1.cpp (hunk 1 above)
Comment (MarekKnapek) [reply Userland/Libraries/LibCrypto/Hash/SHA1.cpp:19]: > macros should be SCREAMING_SNAKE_CASE

OK, fixed. It could be nice if linter could catch this.

File: Userland/Libraries/LibCrypto/Hash/SHA1.cpp (hunk 2 above)
Comment (MarekKnapek) [reply Userland/Libraries/LibCrypto/Hash/SHA1.cpp:23]: > don't need this for non-x86, it's not referenced

OK, will fix that.

File: Userland/Libraries/LibCrypto/Hash/SHA1.cpp (hunk 3 above)
Comment (MarekKnapek) [reply Userland/Libraries/LibCrypto/Hash/SHA1.cpp:34]: > can we use Spans or Array& or something instead of raw pointers with no size information? This is partially pre-existing in this file but there's no reason to propagate that mistake further.

OK, will work on this.

File: Userland/Libraries/LibCrypto/Hash/SHA1.cpp (hunk 4 above)
Comment (MarekKnapek) [reply Userland/Libraries/LibCrypto/Hash/SHA1.cpp:98]: OK, will create variables at theirs first use.

File: Userland/Libraries/LibCrypto/Hash/SHA1.cpp (hunk 5 above)
Comment (MarekKnapek) [reply Userland/Libraries/LibCrypto/Hash/SHA1.cpp:106]: Yes, need to be unrolled as some of the intrinsics / instructions consume only constant integers, not integers from variables.

Comment (MarekKnapek) [reply Userland/Libraries/LibCrypto/Hash/SHA1.cpp:106]: Yes, the spec, NIST FIPS PUB 180-4 is talking about variables with such names. For example in sections 6.2.2.2 and 6.2.2.3.

File: Userland/Libraries/LibCrypto/Hash/SHA1.cpp (hunk 4 above)
Comment (MarekKnapek) [reply Userland/Libraries/LibCrypto/Hash/SHA1.cpp:98]: I would rather not. That would mean having two-element array named `abcd` and four element array `msg`. For the `abcd` that (according to me) does not make sense. The other is doable. If you wish, I will do that. If you insist I will do that `abcd` too.

File: Userland/Libraries/LibCrypto/Hash/SHA1.cpp (hunk 6 above)
Comment (MarekKnapek) [reply Userland/Libraries/LibCrypto/Hash/SHA1.cpp:101]: OK, will do.

File: Userland/Libraries/LibCrypto/Hash/SHA1.cpp (hunk 7 above)
Comment (MarekKnapek) [reply Userland/Libraries/LibCrypto/Hash/SHA1.cpp:259]: No, it is safe. The write goes to variable of type `u32` and the read is a "cast" from (part of) 128bit register. I could improve this by `static_cast`ing the `int` to `u32`.

File: Userland/Libraries/LibCrypto/Hash/SHA2.cpp (hunk 8 above)
Comment (MarekKnapek) [reply Userland/Libraries/LibCrypto/Hash/SHA2.cpp:50]: No, cannot read private SHAxxx::member constants form non-member function. The function needs to be non-member because the attribute target would not work otherwise. Will fix that by moving the constant from header into this function as constexpr variable.

File: Userland/Libraries/LibCrypto/Hash/SHA1.cpp (hunk 2 above)
Comment (MarekKnapek) [reply Userland/Libraries/LibCrypto/Hash/SHA1.cpp]: Fixed.

File: Userland/Libraries/LibCrypto/Hash/SHA1.cpp (hunk 3 above)
Comment (MarekKnapek) [reply Userland/Libraries/LibCrypto/Hash/SHA1.cpp]: Fixed.

File: Userland/Libraries/LibCrypto/Hash/SHA1.cpp (hunk 4 above)
Comment (MarekKnapek) [reply Userland/Libraries/LibCrypto/Hash/SHA1.cpp]: Fixed.

File: Userland/Libraries/LibCrypto/Hash/SHA1.cpp (hunk 6 above)
Comment (MarekKnapek) [reply Userland/Libraries/LibCrypto/Hash/SHA1.cpp]: Fixed.

File: Userland/Libraries/LibCrypto/Hash/SHA1.cpp (hunk 7 above)
Comment (MarekKnapek) [reply Userland/Libraries/LibCrypto/Hash/SHA1.cpp]: Added `static_cast`.

File: Userland/Libraries/LibCrypto/Hash/SHA2.cpp (hunk 8 above)
Comment (MarekKnapek) [reply Userland/Libraries/LibCrypto/Hash/SHA2.cpp]: Fixed.

**Comment by gmta** [comment]
@MarekKnapek 

> Will create benchmark that mimics Serenity code as close as possible. But those numbers would be pulled out of my a**.

How would you even know if your code works correctly and improves performance, then? You should start with being able to actually run your own code in the context of SerenityOS (or Lagom/Ladybird for example) and get performance results from there, because those would be the only relevant benchmarks imho.

**Comment by MarekKnapek** [comment]
Hi @gmta. My situation is complicated. Originally I developed the code for my own project (goal of this project is to implement as much hashing and as much crypto as possible). All my computers are virtual computers. It was done on Windows machine using the MSVC compiler. On virtual computer that does support the SHA instructions. But does not support WSL. It cannot run Serenity. My other virtual computer is a Linux computer that can run Serenity but cannot run the SHA instruction. I'm certain my code is correct because I used the [Intel SDE](https://www.intel.com/content/www/us/en/developer/articles/tool/software-development-emulator.html) simulator, that allows you to test your executable as-if it is running on CPU that supports the SHA instruction.

**Comment by gmta** [comment]
@MarekKnapek 

> My situation is complicated.

Maybe you could reach out on Discord to see if someone with a CPU with the SHA extensions can run these benchmarks for you? That'd at least validate the usage of these instructions.

**Comment by MarekKnapek** [comment]
Benchmark provided by danshaders from discord [link](https://discord.com/channels/830522505605283862/843825120216743947/1229496488179011726):

```c++
#define AK_DONT_REPLACE_STD

#include <chrono>
#include <string>
#include <iostream>

#include <AK/Assertions.h>
#include <LibCrypto/Hash/SHA1.h>

int main() {
    constexpr int n = 5e7;
    constexpr int iterations = 50;

    std::string long_long_string(n, 'a');

    StringView view(long_long_string.data(), n);

    auto start = std::chrono::steady_clock::now();
    for (int i = 0; i < iterations; ++i) {
        AK::taint_for_optimizer(view);
        auto digest = Crypto::Hash::SHA1::hash(view);
        AK::taint_for_optimizer(digest);
    }
    auto end = std::chrono::steady_clock::now();
    std::cout << std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count() << "ms\n";
}
```
```
Clang: without SIMD — 47(50)ms, with SIMD — 114(0)ms, so this is solid 4x speed up
```

**Comment by MarekKnapek** [comment]
After discussion with and great help from @Hendiadyoin1, thank you, I force pushed my branch. It is ready for re-review.

File: Userland/Libraries/LibCrypto/Hash/SHA2.cpp [hunk 9]
```diff
@@ -56,14 +69,67 @@ inline void SHA256::transform(u8 const* data)
...
+    states[0] = AK::SIMD::load_unaligned<i32x4>(&state[0]);
+    states[1] = AK::SIMD::load_unaligned<i32x4>(&state[4]);
+    auto tmp = i32x4 { states[0][1], states[0][0], states[0][3], states[0][2] };
+    states[1] = i32x4 { states[1][3], states[1][2], states[1][1], states[1][0] };
+    states[0] = i32x4 { states[1][2], states[1][3], tmp[0], tmp[1] };
+    states[1] = i32x4 { states[1][0], states[1][1], tmp[2], tmp[3] };
+
+    i32x4 msgs[4] {};
+    i32x4 old[2] { states[0], states[1] };
+    for (int i_msg = 0; i_msg != 16; ++i_msg) {
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibCrypto/Hash/SHA2.cpp:98]: `i` for a name should be enough here

File: Userland/Libraries/LibCrypto/Hash/SHA2.cpp [hunk 10]
```diff
@@ -56,14 +69,67 @@ inline void SHA256::transform(u8 const* data)
...
+        i32x4 msg {};
+        if (i_msg < 4) {
+            msgs[(i_msg + 0) % 4] = AK::SIMD::load_unaligned<i32x4>(&data[i_msg * 16]);
+            msgs[(i_msg + 0) % 4] = AK::SIMD::elementwise_byte_reverse(msgs[(i_msg + 0) % 4]);
+            tmp = AK::SIMD::load_unaligned<i32x4>(&SHA256Constants::RoundConstants[i_msg * 4]);
+            msg = msgs[(i_msg + 0) % 4] + tmp;
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibCrypto/Hash/SHA2.cpp:104]: as this is only executed in the first 4 iterations the `% 4` can be elided

File: Userland/Libraries/LibCrypto/Hash/SHA2.cpp (hunk 9 above)
Comment (MarekKnapek) [reply Userland/Libraries/LibCrypto/Hash/SHA2.cpp:98]: Fixed.

File: Userland/Libraries/LibCrypto/Hash/SHA2.cpp (hunk 10 above)
Comment (MarekKnapek) [reply Userland/Libraries/LibCrypto/Hash/SHA2.cpp:104]: Fixed.

File: AK/SIMDExtras.h [hunk 11]
```diff
@@ -168,30 +183,93 @@ ALWAYS_INLINE static void store4_masked(VectorType v, UnderlyingType* a, Underly
...
+    } else if constexpr (sizeof(Element) == 4) {
+        return T {
+            static_cast<Element>(__builtin_bswap32(static_cast<u32>(a[Idx])))...
+        };
+    } else if constexpr (sizeof(Element) == 8) {
+        return T {
+            static_cast<Element>(__builtin_bswap64(static_cast<u64>(a[Idx])))...
+        };
+    } else {
+        __builtin_unreachable();
```
Comment (DanShaders) [inline AK/SIMDExtras.h:242]:
```suggestion
        static_assert(DependentFalse<T>);
```

File: Userland/Libraries/LibCrypto/Hash/SHA1.cpp [hunk 12]
```diff
@@ -67,6 +81,63 @@ inline void SHA1::transform(u8 const* data)
...
+    i32x4 msgs[4] {};
+
+    abcd[0] = AK::SIMD::load_unaligned<i32x4>(&state[0]);
+    abcd[0] = AK::SIMD::item_reverse(abcd[0]);
+    i32x4 e { 0, 0, 0, static_cast<i32>(state[4]) };
+
+    // FIXME: Test if unrolling the loop is worth it
+    //          GCC: #pragma GCC unroll(5)
+    //        Clang: #pragma unroll
+#    define GROUP(i_group)                                                                         \
```
Comment (DanShaders) [inline Userland/Libraries/LibCrypto/Hash/SHA1.cpp:100]: Why can't this be lambda?

File: Userland/Libraries/LibCrypto/Hash/SHA2.cpp [hunk 13]
```diff
@@ -5,9 +5,19 @@
...
 
+#if (ARCH(I386) || ARCH(X86_64)) && !defined KERNEL && defined(AK_COMPILER_GCC)
+#    include <AK/SIMD.h>
+#    include <AK/SIMDExtras.h>
+#    define SHA256_ATTRIBUTE_TARGET_DEFAULT __attribute__((target("default")))
+#    define SHA256_ATTRIBUTE_TARGET_X86 __attribute__((target("sha"), used))
+#else
+#    define SHA256_ATTRIBUTE_TARGET_DEFAULT
+#endif
+
```
Comment (DanShaders) [inline Userland/Libraries/LibCrypto/Hash/SHA2.cpp:20]: These lines shouldn't be duplicated

File: Userland/Libraries/LibCrypto/Hash/SHA1.cpp (hunk 12 above)
Comment (MarekKnapek) [reply Userland/Libraries/LibCrypto/Hash/SHA1.cpp:100]: Because the `sha1rnds4` instruction takes an `imm8` not variable or pointer or register.

File: Userland/Libraries/LibCrypto/Hash/SHA2.cpp (hunk 13 above)
Comment (MarekKnapek) [reply Userland/Libraries/LibCrypto/Hash/SHA2.cpp:20]: Cannot see any duplication. Can you explain what do you mean? There is conditional definition of `SHA256_ATTRIBUTE_TARGET_DEFAULT` macro, it is needed for non-x86 architectures to have different value.

File: AK/SIMDExtras.h (hunk 11 above)
Comment (MarekKnapek) [reply AK/SIMDExtras.h:242]: OK, will fix.

File: Userland/Libraries/LibCrypto/Hash/SHA1.cpp (hunk 12 above)
Comment (DanShaders) [reply Userland/Libraries/LibCrypto/Hash/SHA1.cpp:100]: Then `i_group` should be a NTTP:
```cpp
auto group = [&]<int i_group>() {
```

File: Userland/Libraries/LibCrypto/Hash/SHA2.cpp (hunk 13 above)
Comment (DanShaders) [reply Userland/Libraries/LibCrypto/Hash/SHA2.cpp:20]: These macros differ from ones in SHA1.cpp only in name.

File: Userland/Libraries/LibCrypto/Hash/SHA1.cpp (hunk 12 above)
Comment (MarekKnapek) [reply Userland/Libraries/LibCrypto/Hash/SHA1.cpp:100]: OK, will fix.

File: Userland/Libraries/LibCrypto/Hash/SHA2.cpp (hunk 13 above)
Comment (MarekKnapek) [reply Userland/Libraries/LibCrypto/Hash/SHA2.cpp:20]: OK, what do you suggest instead? Because this pattern could repeat in the future. Fox example for SHA-512 (when Arrow Lake and Lunar Lake came out) or for AES and VAES. That's what comes to my mind. Maybe also for CRC. And maybe also for different architectures. But each time there will be (possible) different set of conditions.

Comment (DanShaders) [reply Userland/Libraries/LibCrypto/Hash/SHA2.cpp:20]: If I were you, I would add something like
```cpp
#define AK_ATTRIBUTE_TARGET_DEFAULT [[gnu::target("default")]]

// FIXME: Clang doesn't support function multiversioning based on "sha" feature.
#if (ARCH(I386) || ARCH(X86_64)) && defined(AK_COMPILER_GCC)
#    define AK_ATTRIBUTE_TARGET_X86_SHA [[gnu::target("sha"), gnu::used]]
#endif
```
to `SIMD.h` / `SIMDExtras.h` and then guard target-specific definitions of `transform` by `#ifdef AK_ATTRIBUTE_TARGET_X86_SHA`

Comment (DanShaders) [reply Userland/Libraries/LibCrypto/Hash/SHA2.cpp:20]: I've also looked up why clang doesn't support said multiversioning and the situation seems quite hopeless: https://github.com/llvm/llvm-project/issues/59040, https://discourse.llvm.org/t/proposal-split-built-ins-from-the-rest-of-compiler-rt/67978. So, unfortunately, we might want to consider a different approach some time in the future.

Comment (Hendiadyoin1) [reply Userland/Libraries/LibCrypto/Hash/SHA2.cpp:20]: it works on clang trunk though
https://godbolt.org/z/oYa515e9G
(I have mentioned this on discord as well I think)
